src/boot/ZulipAsyncStorage.js

```diff
@@ -18,15 +21,32 @@ export default class CompressedAsyncStorage {
     //   (e.g. "zlib base64")
     // DATA = state in JSON format with TRANSFORMS applied
     // The "z|TRANSFORMS|" part shall be called the "header" of the string.
-    if (result.startsWith(NativeModules.TextCompressionModule.header)) {
-      result = await NativeModules.TextCompressionModule.decompress(result);
+    if (result.startsWith('z')) {
+      const header = result.substring(0, result.indexOf('|', result.indexOf('|') + 1) + 1);
+      if (
+        NativeModules.TextCompressionModule !== undefined
+        && header === NativeModules.TextCompressionModule.header
+      ) {
+        result = await NativeModules.TextCompressionModule.decompress(result);
+      } else {
+        // Panic! If we are confronted with an unknown format, there is
+        // nothing we can do to save the situation. Log an error, clear the
+        // storage and restart the app. This should not be possible unless a
+        // user downgrades their version of the app.
+        logErrorRemotely(new Error(`No decompression module found for format ${header}`), '');
+        AsyncStorage.clear();
+        RNRestart.Restart();
```

Comment:
This else clause turned out to be much harder to implement than expected. Unfortunately, redux-persist doesn't ship some reset-and-rehydrate option (which would've been perfect here). 

In light of this, I went on and tried to return NULL_OBJECT '{}' to reset the state manually this way. But NULL_OBJECT is still an object, and therefore no complete reset. As a result, some reducers that expected arrays failed when confronted with {}. The right thing to do that also happens on the first startup is to return `undefined`, which is a case that our reducers handle appropriately. But `undefined` doesn't exist in JSON, and whatever we return gets parsed by `JSON.parse()`... Same story with custom transformers. One solution to this would be to inspect the requested key and return either {} or [], whatever the null state for that reducer is. But that's hacky and brittle. And since we only expect this situation to occur extremely infrequently, we should just go ahead, reset the storage and do a clean restart. Even better, `RNRestart.Restart();` doesn't do a full restart, but only reloads the JS.

Reply:
Hmm, interesting!

What if we throw an error? (Given the look of this API, I think that'd mean first something like `callback(error, undefined)`, then `throw error`.) That seems like an appropriate response to this situation.

It looks like this API is intended to sometimes throw errors, so I'd hope the callers would handle that reasonably.

Reply:
A possible bit of data here: in my emulator, I had installed this version and then of course ended up "downgrading" (to current master d42189796 plus a short branch) to review a different PR.

I got this error:
![image](https://user-images.githubusercontent.com/28173/43352158-27608188-91d4-11e8-8e91-0f92d5f718d4.png)

That's `null is not an object (evaluating 'users.reduce')`, in this code:
```
export const getUsersById = createSelector(getUsers, (users = []) =>
  users.reduce((usersById, user) => {
    usersById[user.user_id] = user;
    return usersById;
  }, {}),
);
```

And `getUsers` is just `state => state.users`. So we're ending up with a state where `state.users` is `null`. Not sure what causes that, but it's clearly a pretty unreasonable result; better to ignore the persisted state entirely, as if it had disappeared.

Reply:
> What if we throw an error? (Given the look of this API, I think that'd mean first something like callback(error, undefined), then throw error.) That seems like an appropriate response to this situation.

Ah yeah, I looked in the handler code for this:

```js
  let restoredState = {}
  let completionCount = 0

  storage.getAllKeys((err, allKeys) => {
    if (err) {
      if (process.env.NODE_ENV !== 'production') console.warn('redux-persist/getStoredState: Error in storage.getAllKeys')
      complete(err)
    }

    let persistKeys = allKeys.filter((key) => key.indexOf(keyPrefix) === 0).map((key) => key.slice(keyPrefix.length))
    let keysToRestore = persistKeys.filter(passWhitelistBlacklist)

    let restoreCount = keysToRestore.length
    if (restoreCount === 0) complete(null, restoredState)
    keysToRestore.forEach((key) => {
      storage.getItem(createStorageKey(key), (err, serialized) => {
        if (err && process.env.NODE_ENV !== 'production') console.warn('redux-persist/getStoredState: Error restoring data for key:', key, err)
        else restoredState[key] = rehydrate(key, serialized)
        completionCount += 1
        if (completionCount === restoreCount) complete(null, restoredState)
      })
    })
  })
```

So in case of an error, redux-persist behaves in two different ways:

* if we are in development mode and receive an error, print a warning and restore the state `{}`.

* if we are in production mode, proceed as if the warning didn't exist.

Reply:
Both scenarios are undesirable for us: Just using `{}` doesn't work, (as explained in the comment above). And pretending that there is no error of course doesn't work either.

Reply:
> better to ignore the persisted state entirely, as if it had disappeared.

So that's what I meant with
>  One solution to this would be to inspect the requested key and return either {} or [], whatever the null state for that reducer is. But that's hacky and brittle.

This is brittle because we'd just ignore the persisted state for a single key. But say for some reason, we ignore it for `messages` but manage a rehydration for `caughtUp`, then we might end up thinking we have messages that we don't actually have. So it'd indeed be best to just ignore the **entire** state, but that's difficult with redux-persist's key-based interface.

Reply:
Thus, I was thinking that the best solution is to just `AsyncStorage.clear()` the whole thing and then restart the app. The error you're getting should not pop up in production, right? I was hoping that in production, the app would just silently restart.

Reply:
That said, I don't understand why the error pops up even though we already initiated the restart...

Reply:
Oh, wait, I think I now know the reason why you are getting this annoying error message: When you say you "downgraded" to master, it means that you actually are downgrading to a version that is completely unaware of compression, right? In fact, it is so unaware that it doesn't even know that the stored state could be compressed, and hence it doesn't know how to handle that situation. In such a situation, I think we can't do any better than this. We cannot change the code of already installed apps (except perhaps with "updates on the fly"? but that's a different story.)

The best we can do is to handle error from now on. You can test a  "downgrade", by commenting out 

```java
File: android/app/src/main/java/com/zulipmobile/ZulipNativePackage.java
29:         modules.add(new TextCompressionModule(reactContext));
```

Reply:
Huh wait interesting:

> storage.getAllKeys

So, it looks like the `redux-persist` code you're looking at is operating on the *root* state -- not on a single property / subtree of that state. That is, it's operating on `accounts`, `drafts`, `messages`, `users`, etc., etc., all at once, not on any one of them.

Which means that `{}` is exactly the right answer if we have no data.

More generally, though: somehow this system does something reasonable if there really is no data to be had; so there had to be a way to get that result, whatever it is.

Reply:
Still more interesting, though:

> storage.getAllKeys

That is a method we're not overriding -- instead we're just carrying it through from `AsyncStorage`.

I took at look at the implementation of that method in `AsyncStorage`. It doesn't use the `getItem` method! Instead it goes straight to the Java backend:
```
  getAllKeys: function(callback?: ?(error: ?Error, keys: ?Array<string>) => void): Promise {
    return new Promise((resolve, reject) => {
      RCTAsyncStorage.getAllKeys(function(error, keys) {
        callback && callback(convertError(error), keys);
        if (error) {
          reject(convertError(error));
        } else {
          resolve(keys);
        }
      });
    });
  },
```

So, it seems like this ought to bypass the part where we so carefully decompress. Perhaps we should rewrite `getAllKeys`?

(This is also making me think maybe it would all be easier if we did the wrapping in Java anyway. `ZulipAsyncStorage` has very little logic of its own, so it's not like this would mean writing a ton of Java code.)

Reply:
OK, I think I found a solution. I am not impressed with redux-persist's treatment of errors here -- that thing where when it gets an error, it barrels on ahead with attempting to process whatever garbage it thinks it got (and only in production, too!) is really something -- but there's at least one important thing it gets right here: if the result for a given key is `null`, then the built-in rehydrate logic seems to handle that properly, so the state we wind up with under that key remains `{}` or `[]` or whatever it was that we'd initialized it to.

We just need to handle it properly in the few spots where we have custom rehydrate logic of our own.

One of those is `navReducers` -- that's the one where a failure means we get stuck on the loading screen forever. I tried fixing that, and now if we just raise an error in the normal way, everything degrades gracefully.

Reply:
Aah, great!